libs/ml/examples/cbow_simple/main.cpp

```diff
@@ -0,0 +1,333 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include <math.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <chrono>
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+
+#include "math/approx_exp.hpp"
+#include "math/tensor.hpp"
+#include "polyfill.hpp"
+#include "unigram_table.hpp"
+#include "w2v_cbow_dataloader.hpp"
+#include "word_loader.hpp"
+
+using namespace std::chrono;
+using namespace fetch::ml;
+using FloatType = float;
+
+CBOWLoader<FloatType> data_loader(5, 25);
+WordLoader<FloatType> new_loader;
+
+std::string train_file, output_file;
+
+int32_t window    = 5;  // 5
+int32_t min_count = 5;
+
+double time_forward  = 0;
+double time_exp      = 0;
+double time_backward = 0;
+double time_step     = 0;
+
+uint64_t                          dimensionality = 200;  // 200
+uint64_t                          iter           = 1;
+FloatType                         alpha          = static_cast<FloatType>(0.025);
+FloatType                         starting_alpha;
+int32_t                           negative = 25;
+high_resolution_clock::time_point last_time;
+
+std::string readFile(std::string const &path)
+{
+  std::ifstream t(path);
+  return std::string((std::istreambuf_iterator<char>(t)), std::istreambuf_iterator<char>());
+}
+
+void PrintStats(uint32_t const &i, uint32_t const &iterations)
```

Comment:
this does more than just print stats! it updates the learning rate in a linear step decay